### ISSUE TITLE
Nested LUA coroutines cause crashes on MACs

### ISSUE THREAD

**Issue by wookieejedi** [scp-fs2open/fs2open.github.com#5904]
I found that nesting LUA coroutines causes crashes on MAC. Overall single coroutines work just fine on Mac, but if I make a coroutine and then call a function that runs another coroutine the crash occurs (again only on MAC). The error is thrown from `LuaFunction.cpp:155` and states `Error: Reference does not refer to a function!` (an image of this error and using the provided simple repo case is attached below).

![image](https://github.com/scp-fs2open/fs2open.github.com/assets/14077810/2ab1b282-cbce-457e-99d0-82f44935cc20)

Me and taylor and chief were trying to figure out how to fix it, but alas we came up short, and taylor suggested asking someone with more experience with FSO LUA internals. 

I've attached a simple repro case, which is just a single small script that will trigger when any mission starts. I've also attached an image of the code for convenience. 
[corotest-sct.zip](https://github.com/scp-fs2open/fs2open.github.com/files/13795540/corotest-sct.zip)

![image](https://github.com/scp-fs2open/fs2open.github.com/assets/14077810/194705fa-55b6-4e83-b1ee-a9078b85a64a)

Wondering if either @BMagnu or @asarium might have any insight on how to fix this, especially since it is preventing MAC users from using certain scripts and playing FotG.

**Comment by asarium** [comment]
Please also add a stacktrace of the process when this error occurs to determine which exact code path is causing the issue here.

**Comment by wookieejedi** [comment]
This is from chief yesterday though I can ask him or taylor for more logs or a different trace if that's useful?
```
ok so the crash wasn't 
automatically caught by the debugger, but since we knew where Lua triggered the throw(), I was able to put a manual breakpoint there, and now I have it stopped at that point
```
[debug_report.txt](https://github.com/scp-fs2open/fs2open.github.com/files/13795598/debug_report.txt)

@notimaginative, perhaps you might have a stracktrace?

**Comment by asarium** [comment]
The debug report contained a stacktrace so that is good enough.

The weird thing about the stack trace is that in `luacpp::convert::popValue` there is a check for the value being a Lua function (https://github.com/scp-fs2open/fs2open.github.com/blob/master/code/scripting/lua/LuaFunction.cpp#L240) which obviously succeeds.

However, the check that fails (https://github.com/scp-fs2open/fs2open.github.com/blob/master/code/scripting/lua/LuaFunction.cpp#L153) checks for exactly the same thing so that is weird and would be a good place to start the investigation.

The only difference between the two checks is that the second variant uses the main Lua state/thread which is what gets stored in the Lua reference so that difference could be significant here.

**Comment by notimaginative** [comment]
I'm not sure if it will help or push this too far to the right of wtf, but I originally encountered this crash several years ago on Linux and was never able to figure it out. However, for some reason, the crash only happened when I compiled FSO with clang. If GCC was used instead then everything worked as expected. Using the included lua lib or a system version made no difference.

The Mac compiler toolkit is based on clang, which may be why this showed up again, and only on Mac.

**Comment by asarium** [comment]
If this is actually down to a compiler issue then I do not think that I can help any further. I looked through the code a bit again and could not find anything that would either directly cause issues or run into some undefined behavior which might explain why Clang could behave differently.

Except for direct debugging of what exactly is happening here, there is probably not much else we can do at this point.

**Comment by notimaginative** [comment]
Which is unfortunately where we ended up last time. I tried debugging it previously but I just don't know enough about the lua code to understand what's going on and what to look for.

At least it's not Mac specific, so someone on Linux with enough lua knowledge (or time) might could track it down. And I've never tried it but I believe clang can be installed and used with MSVC ("clang tools for windows", or something like that, in the installer). So maybe if this could be reproduced on Windows, with MSVC, it would help with debugging.
